tools/hdf_dump_translation_dataset.py

```diff
@@ -0,0 +1,322 @@
+
+    self.hdf_file.create_dataset("seqLengths", (self.number_of_lines, 2), dtype="int32")
+    max_tag_length = len("line-") + len(str(self.number_of_lines))
+    self._tag_dtype = "S{}".format(max_tag_length)
+    self.hdf_file.create_dataset("seqTags", (self.number_of_lines,), dtype=self._tag_dtype)
+
+    self.hdf_file.create_dataset("inputs", (self.data_buffer_size,), maxshape=(None,),
+                                 dtype="int32", compression=self.compression_method)
+    self.hdf_file["targets/data"].create_dataset("classes", (self.data_buffer_size,), maxshape=(None,),
+                                                 dtype="int32", compression=self.compression_method)
+
+  def _write_attributes(self):
+    """
+    Writes several attributes to the HDF file.
+    """
+    self.hdf_file["targets/size"].attrs["classes"] = (self._vocabulary_sizes["target"], 1)
+
+    # Those should be deprecated, but include nevertheless to exactly reproduce hdf_dump.
+    self.hdf_file.attrs["inputPattSize"] = self._vocabulary_sizes["source"]
+    self.hdf_file.attrs["numLabels"] = self._vocabulary_sizes["target"]
+
+  def _write_labels(self):
+    """
+    Writes the labels (i.e. target vocabulary) to the HDF file.
+    """
+    sorted_vocabulary_tuples = sorted(self._vocabularies["target"].items(), key=itemgetter(1))
+    labels = [word.encode("utf8") for (word, _) in sorted_vocabulary_tuples]
+
+    assert len(labels) == self._vocabulary_sizes["target"], "Word ids were not unique."
+
+    max_label_length = max([len(label) for label in labels])
+    dtype = "S{}".format(max_label_length + 1)
+
+    labels = [numpy.array(label, dtype=dtype, ndmin=1) for label in labels]
+    labels = numpy.concatenate(labels)
+
+    self.hdf_file["targets/labels"].create_dataset("classes", (self._vocabulary_sizes["target"],),
+                                                   data=labels, dtype=dtype)
+
+  def _write_data(self):
+    """
+    Loads a chunk of lines from the corpus and writes all corresponding data to the HDF file.
+
+    :return: whether the end of the corpus is reached
+    :rtype: bool
+    """
+    source_chunk, target_chunk = self._get_chunk()
+
+    if not source_chunk:
+      self._finalize_data()
+      return True
+
+    source_lengths = [len(line) for line in source_chunk]
+    target_lengths = [len(line) for line in target_chunk]
+
+    chunk_size = len(source_chunk)
+
+    self._write_lengths(source_lengths, target_lengths)
+    self._write_tags(chunk_size)
+
+    self._write_data_indices(source_chunk, "source")
+    self._write_data_indices(target_chunk, "target")
+
+    self._number_of_processed_lines += chunk_size
+
+    return False
+
+  def _get_chunk(self):
+    """
+    Reads in the next chunk of lines from the corpus files.
+
+    :return: Word indices for all lines in the current chunk, for both source and target.
+        Indices for one line are stored as a numpy array of dtype int and shape (None,).
```

Comment:
Is there a standard syntax for defining the array attributes?

Reply:
I think the PEP defines this partly (or at least suggests ways).
PyCharm has some documentation about that (which is maybe a bit extended from PEP).
`:rtype:` is just about the Python type (what you would get when you call `type(obj)`). The Python type is `numpy.ndarray` here (I guess). You can also use the `typing` module, and some PEP/PyCharm extended syntax (e.g. `dict[int,str]` works here; but not in other situations).

Reply:
It would be `:rtype: numpy.ndarray` here. All other info in `:return:`.

Reply:
Not sure I understand what to change now. The type() of the returned object would just be tuple...
I meant to ask whether the way I described the dtype and shape is ok.

Reply:
Ah, sorry, misunderstood.
But you don't need to say that it is a numpy array, because that can already be seen in `:rtype:`. Keep it short. That makes it easier to read and comprehend (I think). I would write it maybe just like:
```
:return: word indices for source and target, dtype int, shape [Lines]
:rtype: list[numpy.ndarray], list[numpy.ndarray]
```
You already mention that it is for the lines in the sentence just before, same for chunk. Don't repeat, it just blows it up unnecessarily and makes it harder to read (in my opinion).

Reply:
Ok, done, I think.